Refactor xk6-browser front page documentation (#731)
grafana/k6-docs: src/data/markdown/docs/30 xk6-browser/30 xk6-browser.md

import launcher from 'k6/x/browser';

export default function () {
  const browser = launcher.launch('chromium', {
    headless: false,

>   });

Reviewer: WDYT about moving this example to the "Example" section, and replacing the example that's currently there? The one there isn't very useful, and we could merge the options demo with this one.

The main reason is because we should first explain what xk6-browser is and how to use/build it, before showing any examples. Otherwise people might try to run the example without reading the rest of the required steps.

BTW, remember to change it to use the new API from https://github.com/grafana/xk6-browser/pull/462.

Author: Sure, i'll make that change.

I'll amend the example in a different PR to conform to the API change in https://github.com/grafana/xk6-browser/pull/462.

Author: Partially resolved in 43f0d9825beaf62ff70739bc52a0859cf60d1524

<InstallationInstructions extensionUrl="github.com/grafana/xk6-browser"/>

## Launch

> The first step is to launch a [Browser](/javascript-api/xk6-browser/browser). After a browser process starts up you will be able to interact with it using the [browser-level APIs](#browser-level-apis) listed below. Currently `chromium` is the only accepted browser type.

Reviewer 2: ```suggestion
The first step is to launch a [Browser](/javascript-api/xk6-browser/browser). After the browser starts, you can interact with it using the [browser-level APIs](#browser-level-apis). The only accepted browser type is currently `chromium`.
```

| env               | Environment variables to set before launching browser process.                                                                                                                      |
| executablePath    | Override search for browser executable in favor of specified absolute path.                                                                                                         |
| headless          | Show browser GUI or not. Defaults to `true`.                                                                                                                                        |
| ignoreDefaultArgs | Ignore any of the default arguments included when launching browser process.                                                                                                        |
| proxy             | Specify to set browser's proxy config.                                                                                                                                              |

> | slowMo            | Slow down input actions and navigation by specified time. By default this is off.                                                                                                   |

Reviewer 2: ```suggestion
| slowMo            | Slow down input actions and navigation by the specified time. By default, it is off.                                                                                                 |
```

| timeout           | Default timeout to use for various actions and navigation. Default timeout is 30 seconds.                                                                                           |

### Example

<CodeGroup labels={[]}>

<!-- eslint-skip -->

```javascript
import launcher from 'k6/x/browser';

export default function () {
  const browser = launcher.launch('chromium', {

>     args: [],

Reviewer 2: Since it's empty, we can omit it.

Author: I'll try to find a suitable arg to place in there

    debug: true,
    devtools: true,

>     env: {},

Reviewer 2: We can also omit this.

Author: And I'll add an env var here too along with how it can be used in the test

>     executablePath: null,

Reviewer 2: We can omit this.

Author: I'll add a suitable path here

    headless: false,

>     ignoreDefaultArgs: [],

Reviewer 2: We can omit this.

>     proxy: {},

Reviewer 2: We can omit this.

    slowMo: '500ms,

>     timeout: '30s,

Reviewer 2: ```suggestion
    timeout: '30s',
```

  });

>   // perform tests against your website...

Reviewer 2: ```suggestion
  // add your test code here and finally close the browser
```

<InstallationInstructions extensionUrl="github.com/grafana/xk6-browser"/>

## Launch

> The first step is to launch a [Browser](/javascript-api/xk6-browser/browser). After the browser starts, you can interact with it using the [browser-level APIs](#browser-level-apis). The only accepted browser type is currently `chromium`.

Reviewer: Since we changed the API in https://github.com/grafana/xk6-browser/pull/462, let's update the last sentence here, and the `launch()` documentation below.

Actually, since we now export browser names directly, it might make sense to have separate browser sections here, instead of the "Launch" `h2`. This would make it clear that only `chromium` is currently supported, and also allow us to expand the docs once support for other browsers is added.

Author: Sounds good to me. Can i do these changes in a separate PR? I'm going to work on the docs for the https://github.com/grafana/xk6-browser/pull/462 change in one single PR.

|--------------------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| launch(browserName, [options]) | The first parameter is required, and the only accepted `browserName` is currently `chromium`. The `options` parameter is optional. You can see a list of all possible `options` in [the options heading](#options). |

### Options

| Method            | Description                                                                                                                                                                                              |
|-------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|

> | args              | Extra command line arguments to include when launching browser process. See [this link](https://peter.sh/experiments/chromium-command-line-switches/) for a list of Chromium arguments. Note that arguments should not start with `--` (see the [example](#Example)).  |

Reviewer: I wish there was an official list somewhere instead of some random blog. I wasn't able to find another source though :-/

| debug             | All CDP messages and internal fine grained logs will be logged if set to `true`. Default is `false`.                                                                                                     |

> | devtools          | Open up developer tools in the browser by default. Default is `false`.                                                                                                                                   |

Reviewer: WDYT about adding a "Default" column? [We do this on some pages](https://k6.io/docs/using-k6/scenarios/executors/ramping-vus/) (though I would place the column after "Type" instead), and it makes it easy to see at a glance.

Author: That makes sense. I might do that in another PR though so that I can add default columns to all tables i've added/amended.

Author: Issue created: https://github.com/grafana/xk6-browser/issues/473

Author: Resolved in: c7fafbff787583dc59e4799d1bd4c5c21ef69a78

|--------------------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| launch(browserName, [options]) | The first parameter is required, and the only accepted `browserName` is currently `chromium`. The `options` parameter is optional. You can see a list of all possible `options` in [the options heading](#options). |

### Options

| Method            | Description                                                                                                                                                                                              |
|-------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|

> | args              | Extra command line arguments to include when launching browser process. See [this link](https://peter.sh/experiments/chromium-command-line-switches/) for a list of Chromium arguments; remove the `--`. |

Reviewer 2: ?

```suggestion
| args              | Extra command line arguments to include when launching browser process. See [this link](https://peter.sh/experiments/chromium-command-line-switches/) for a list of Chromium arguments. Note that arguments should not start with `--`.  |
```

| Method                          | Description                                                                                                                                                                                 |
|---------------------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|

> | launch(browserType, [options]) | The first argument is required and the only `browserType` that is currently accepted is `chromium`. Below you will find a list of all possible `options`. The `options` param is optional. |

Reviewer 2: ```suggestion
| launch(browserName, [options]) | The first parameter is required, and the only accepted `browserName` is currently `chromium`. The `options` parameter is optional. You can see a list of all possible `options` in the options heading.
```

By saying the "options heading" instead of "below", we'd be invincible even if the layout changes.

Author: Yep, all sounds good 👍

| [Response](/javascript-api/xk6-browser/response/) | [`finished()`](https://playwright.dev/docs/api/class-response#response-finished) (dependent on event-loop support in k6) |
| [Touchscreen](/javascript-api/xk6-browser/touchscreen/) |  |

The following Playwright APIs are not supported yet:

<Glossary>

- [Accessibility](https://playwright.dev/docs/api/class-accessibility)
- [BrowserServer](https://playwright.dev/docs/api/class-browserserver)
- [CDPSession](https://playwright.dev/docs/api/class-cdpsession)
- [ConsoleMessage](https://playwright.dev/docs/api/class-consolemessage)
- [Coverage](https://playwright.dev/docs/api/class-coverage)
- [Dialog](https://playwright.dev/docs/api/class-dialog)
- [Download](https://playwright.dev/docs/api/class-download)
- [FetchRequest](https://playwright.dev/docs/api/class-fetchrequest)
- [FetchResponse](https://playwright.dev/docs/api/class-fetchresponse)
- [FileChooser](https://playwright.dev/docs/api/class-filechooser)

> - [Locator](https://playwright.dev/docs/api/class-locator)

Reviewer 2: We have the Locator API to some extent. See https://github.com/grafana/xk6-browser/issues/100.

Author: Yeah, i'm working on locator docs in a separate task: https://github.com/grafana/xk6-browser/issues/459 -- PR will be ready next week.

## Browser-level APIs

`xk6-browser` uses [Chrome DevTools Protocol](https://chromedevtools.github.io/devtools-protocol/) (CDP) to instrument and interact with the browser. The `xk6-browser` APIs aims for rough compatibility with the [Playwright API for NodeJS](https://playwright.dev/docs/api/class-playwright). 

Note that because k6 does not run in NodeJS, `xk6-browser` APIs will slightly differ from their Playwright counterparts.

| k6 Class |  Description |
| - |  - |

> | <BWIPT /> [Browser](/javascript-api/xk6-browser/browser/) | It is the entry point for all tests. It is used to launch `BrowserContext`s and `Page`s. |

Reviewer 2: ```suggestion
| <BWIPT /> [Browser](/javascript-api/xk6-browser/browser/) | The entry point for all tests and used to launch `BrowserContext`s and `Page`s. |
```

> | <BWIPT /> [BrowserContext](/javascript-api/xk6-browser/browsercontext/) | It enables independent browser sessions, with separate `page`s, cache and cookies. |

Reviewer 2: ```suggestion
| <BWIPT /> [BrowserContext](/javascript-api/xk6-browser/browsercontext/) | Enables independent browser sessions with separate [Page](/javascript-api/xk6-browser/page/)s, cache, and cookies. |
```

| <BWIPT /> [ElementHandle](/javascript-api/xk6-browser/elementhandle/) | Represents an in-page DOM element. |
| <BWIPT /> [Frame](/javascript-api/xk6-browser/frame/) | Access and interact with the `Page`'s `Frame`s. |
| [JSHandle](/javascript-api/xk6-browser/jshandle) | Represents an in-page JavaScript object. |
| [Keyboard](/javascript-api/xk6-browser/keyboard/) | Used to simulate the keyboard interactions with the associated `Page`. |
| [Mouse](/javascript-api/xk6-browser/mouse/) | Used to simulate the mouse interactions with the associated `Page`. |
| <BWIPT /> [Page](/javascript-api/xk6-browser/page/) | `Page` provides methods to interact with a single tab in a `Browser`. |
| <BWIPT /> [Request](/javascript-api/xk6-browser/request/) | Used to keep track of the request the `Page` makes.  |
| <BWIPT /> [Response](/javascript-api/xk6-browser/response/) | Represents the response received by the `Page`. |

> | [Touchscreen](/javascript-api/xk6-browser/touchscreen/) | Used to simulate touch interactions with the associated `Page`. |

Reviewer 2: We can provide links to these as well: Browser, Page, and Frame.

Author: The table contains links to these classes already. Is it going to be confusing if we add the same links in the description?

Reviewer 2: The one I see that might not have a link is the following:

```
`Page` provides methods to interact with a single tab in a `Browser`.
```

 `Page` might not be linked because it already talks about `Page`, however, `Browser` might have a link.

In there, we could also remove `Page` as follows:

```
Provides methods to interact with a single tab in a [`Browser`](/javascript-api/xk6-browser/browser/).
```

However, for `Keyboard`, for example. It talks about `Page`. So `Page` can be a link without any confusion.

Here is another example, this one can simply say "frame" instead of "`Frame`":

```
| <BWIPT /> [Frame](/javascript-api/xk6-browser/frame/) | Access and interact with the `Page`'s frames. |
```

Author: Ok, sounds good.

Can there be more than one frame in a page?
